@cavvia => Grouped search results lab feature

mzikherman opened the pull request:
Having this lab feature turned on supersedes the spotlight AB test, and could be safe to merge/deploy in the meantime (while the AB test is running), if we're happy with the approach and want to QA further.

![bp](https://user-images.githubusercontent.com/1457859/40740710-8ac46900-6417-11e8-8aca-a8c00c5a5ab3.gif)


The idea is that we flatten out the hash returned from Gravity, and the _first_ element in each group, gets an additional property which is the category itself. Then, the dynamic template for each item in the dropdown knows to render a top line of category when this is present, which results in quite decent scrolling/selecting behavior.

mzikherman commented on src/desktop/components/main_layout/templates/index.jade:
```diff
@@ -9,7 +9,7 @@ html(
   data-user-type=sd.CURRENT_USER && sd.CURRENT_USER.type
   data-lab-features=sd.CURRENT_USER && sd.CURRENT_USER.lab_features && sd.CURRENT_USER.lab_features.join(',')
-  data-autosuggest-search=sd.AUTOSUGGEST_SEARCH
+  data-autosuggest-search=!(sd.CURRENT_USER && sd.CURRENT_USER.lab_features && sd.CURRENT_USER.lab_features.indexOf('Grouped Search Results') != -1) && sd.AUTOSUGGEST_SEARCH === 'experiment'
```
We want to add some custom CSS if youre in the experiment spotlight test group..._and dont have this new lab on_

cavvia replied:
this is starting to get uncomfortably complex

mzikherman replied:
yea...i dont think its _that_ bad, but its getting rough.

mzikherman commented on src/desktop/components/search_bar/collections/search.coffee:
```diff
@@ -23,21 +26,34 @@ module.exports = class Search
+      Object.keys(items).forEach((key) ->
+        if items[key].length > 0
+          firstItem = items[key].shift()
+          firstItem.displayHeading = true
```
The first item in each category gets these additional properties, so the template knows to render the category 'heading'.

cavvia replied:
:+1:

mzikherman commented on src/desktop/components/search_bar/collections/search.coffee:
```diff
@@ -23,21 +26,34 @@ module.exports = class Search
+  parseResults: (items, aggregations) ->
+    if aggregations
+      results = []
+      Object.keys(items).forEach((key) ->
```
This doesn't put the groupings in any particular order, if we wanted to always have 'Artists', then 'Sales', etc. we could easily do that here.

cavvia replied:
Yes we need to enforce the order outlined here: https://www.notion.so/artsy/Discover-c25acaaf8e6d4b9480a66826bd521bf1?p=e7dd56c8ccb84b05a4f048202a2dc750 

Shows are not in auto-suggest you can ignore that type.

mzikherman replied:
Cool, wasn't sure from that doc if the order was decided or not. I'll add the ordering, and thank you for the gallery/institution breakdown as well!

cavvia commented on src/desktop/components/search_bar/collections/search.coffee:
```diff
@@ -23,21 +26,34 @@ module.exports = class Search
+        if items[key].length > 0
+          firstItem = items[key].shift()
+          firstItem.displayHeading = true
+          firstItem.category = _s.capitalize(key)
```
according to the comp we need to pluralize here as well - see https://artsyproduct.atlassian.net/browse/DISCOVER-112

mzikherman replied:
👍

cavvia commented:
Related type PR on the Gravity side https://github.com/artsy/gravity/pull/11757

mzikherman commented:
@cavvia I'll merge so we can check this out on staging. I addressed your comments: pluralizing headers and enforcing an order to some of the categories (and making those categories first).
